modules/3d/test/ocl/test_tsdf.cpp

```diff
@@ -571,6 +571,282 @@ void valid_points_test_common_framesize(VolumeType volumeType, VolumeTestSrcType
+    Matx44f pose;
+    vsRot.getVolumePose(pose);
+    pose = Affine3f(Vec3f(1, 1, 1), Vec3f()).matrix;
+    vsRot.setVolumePose(pose);
+    Volume volumeRot(VolumeType::HashTSDF, vsRot);
+
+    Size frameSize(vs.getRaycastWidth(), vs.getRaycastHeight());
+    Matx33f intr;
+    vs.getCameraIntegrateIntrinsics(intr);
+    bool onlySemisphere = false;
+    float depthFactor = vs.getDepthFactor();
+    Vec3f lightPose = Vec3f::all(0.f);
+    Ptr<Scene> scene = Scene::create(frameSize, intr, depthFactor, onlySemisphere);
+    std::vector<Affine3f> poses = scene->getPoses();
+
+    Mat depth = scene->depth(poses[0]);
+    UMat udepth;
+    depth.copyTo(udepth);
+
+    volume0.integrate(udepth, poses[0].matrix);
+    volumeRot.integrate(udepth, poses[0].matrix);
+
+    UMat upts, unrm, uptsRot, unrmRot;
+
+    volume0.raycast(poses[0].matrix, upts, unrm);
+    volumeRot.raycast(poses[0].matrix, uptsRot, unrmRot);
+
+    Mat mpts = upts.getMat(ACCESS_READ), mnrm = unrm.getMat(ACCESS_READ);
+    Mat mptsRot = uptsRot.getMat(ACCESS_READ), mnrmRot = unrmRot.getMat(ACCESS_READ);
+
+    if (cvtest::debugLevel > 0)
+    {
+        displayImage(depth, mpts, mnrm, depthFactor, lightPose);
+        displayImage(depth, mptsRot, mnrmRot, depthFactor, lightPose);
+    }
+
+    std::vector<Mat> ptsCh(3), ptsRotCh(3);
+    split(mpts, ptsCh);
+    split(uptsRot, ptsRotCh);
+    Mat maskPts0 = ptsCh[2] > 0;
+    Mat maskPtsRot = ptsRotCh[2] > 0;
+    Mat maskNrm0 = nanMask(mnrm);
+    Mat maskNrmRot = nanMask(mnrmRot);
+    Mat maskPtsDiff, maskNrmDiff;
+    cv::bitwise_xor(maskPts0, maskPtsRot, maskPtsDiff);
+    cv::bitwise_xor(maskNrm0, maskNrmRot, maskNrmDiff);
+    double ptsDiffNorm = cv::sum(maskPtsDiff)[0]/255.0;
+    double nrmDiffNorm = cv::sum(maskNrmDiff)[0]/255.0;
+
+    EXPECT_LE(ptsDiffNorm, 786);
+    EXPECT_LE(nrmDiffNorm, 786);
+
+    double normPts = cv::norm(mpts, mptsRot, NORM_INF, (maskPts0 & maskPtsRot));
+    Mat absdot = normalsError(mnrm, mnrmRot);
+    double normNrm = cv::norm(absdot, NORM_L2, (maskNrm0 & maskNrmRot));
+
+    EXPECT_LE(normPts, 2.0);
+    EXPECT_LE(normNrm, 73.05);
+}
+
+
+void debugVolumeDraw(const Volume &volume, Affine3f pose, Mat depth, float depthFactor, std::string objFname)
+{
+    Vec3f lightPose = Vec3f::all(0.f);
+    UMat points, normals;
+    volume.raycast(pose.matrix, points, normals);
+
+    UMat ptsList, ptsList3, nrmList, nrmList3;
+    volume.fetchPointsNormals(ptsList, nrmList);
+    // transform 4 channels to 3 channels
+    cvtColor(ptsList, ptsList3, COLOR_BGRA2BGR);
+    cvtColor(ptsList, nrmList3, COLOR_BGRA2BGR);
+    savePointCloud(objFname, ptsList3, nrmList3);
+
+    displayImage(depth, points.getMat(ACCESS_READ), normals.getMat(ACCESS_READ), depthFactor, lightPose);
+}
+
+
+void boundingBoxGrowthTest(VolumeType volumeType)
+{
+    VolumeSettings vs(volumeType);
+    Volume volume(volumeType, vs);
+
+    if ( volumeType == VolumeType::TSDF || volumeType == VolumeType::ColorTSDF )
```

Comment:
I propose to split the function into separate tests as soon as if-else branches do not have common logic:
- Parameterized GTest test for TSDF and colorTSDF. It's very small and simple.
- Parameterized test with bool setEnableGrowth(true/false). It's for else branch and it'll be smaller than the original one without code duplication.